### ISSUE TITLE
Building for windows 

### ISSUE THREAD

**Issue by ashishmagar600** [Dobiasd/frugally-deep#203]
Thank you for sharing your work. It's been very helpful to me in using Keras based ML models with CPP/FORTRAN work environments (in Linux). 

But I want to build my code as an exe for windows, any suggestions how do I proceed?

**Comment by Dobiasd** [comment]
Using visual C++ might be the most popular way to write/compile C++ code on Windows. I'd suggest to check this one out.

**Comment by ashishmagar600** [comment]
Thanks. But I am trying to build a basic CPP program from Linux to be used on windows. I figured out to use mingw-w64 to compile (from Linux; w/o using fdeep) and run the same program (on windows).

However, if I include fdeep, I have to manually include the directories (using the flag -I), and still, there is a huge error. Don't know if I am doing something wrong.

**Comment by Dobiasd** [comment]
> mingw-w64 to compile (from Linux; w/o using fdeep)

mingw-w64 is GCC for Windows, not a cross compiler to create Windows binaries on a Linux system, I think.

> there is a huge error. Don't know if I am doing something wrong.

Hard to tell without you showing the error messages. :wink:

First guess: Make sure the compiler supports C++14.

**Comment by ashishmagar600** [comment]
I am starting from scratch. So I compiled a "hello world" exe from mingw-w64 on Linux and ran on windows cmd. It worked.
Same Hello World code I just added the fdeep header, and some basic code to read a model and predict. Here is the code and the error log.

[code.txt](https://github.com/Dobiasd/frugally-deep/files/4089370/code.txt)
[log_error.txt](https://github.com/Dobiasd/frugally-deep/files/4089371/log_error.txt)


Thanks.

**Comment by Dobiasd** [comment]
OK, it seems like the compiler you are using does not properly support C++14 (or it's not enabled).

---

1) `error: ‘mutex’ is not a member of ‘std’` (check for thread support):

Please make sure you can compile and run the following example code snippet

```cpp
#include <iostream>
#include <map>
#include <string>
#include <chrono>
#include <thread>
#include <mutex>
 
std::map<std::string, std::string> g_pages;
std::mutex g_pages_mutex;
 
void save_page(const std::string &url)
{
    // simulate a long page fetch
    std::this_thread::sleep_for(std::chrono::seconds(2));
    std::string result = "fake content";
 
    std::lock_guard<std::mutex> guard(g_pages_mutex);
    g_pages[url] = result;
}
 
int main() 
{
    std::thread t1(save_page, "http://foo");
    std::thread t2(save_page, "http://bar");
    t1.join();
    t2.join();
 
    // safe to access g_pages without lock now, as the threads are joined
    for (const auto &pair : g_pages) {
        std::cout << pair.first << " => " << pair.second << '\n';
    }
}
```

source: https://en.cppreference.com/w/cpp/thread/mutex

---

2) `template argument deduction/substitution failed`:

Please make sure you can compile and run the following example code snippet (check for C++14):

```cpp
#include <vector>
#include <iostream>
#include <algorithm>
#include <functional>
 
int main()
{
    std::vector<int> c = {1, 2, 3, 4, 5, 6, 7};
    int x = 5;
    c.erase(std::remove_if(c.begin(), c.end(), [x](int n) { return n < x; }), c.end());
 
    std::cout << "c: ";
    std::for_each(c.begin(), c.end(), [](int i){ std::cout << i << ' '; });
    std::cout << '\n';
 
    // the type of a closure cannot be named, but can be inferred with auto
    // since C++14, lambda could own default arguments
    auto func1 = [](int i = 6) { return i + 4; };
    std::cout << "func1: " << func1() << '\n';
 
    // like all callable objects, closures can be captured in std::function
    // (this may incur unnecessary overhead)
    std::function<int(int)> func2 = [](int i) { return i + 4; };
    std::cout << "func2: " << func2(6) << '\n';
}
```

source: https://en.cppreference.com/w/cpp/language/lambda

---

Once those two are working, you can try again with frugally-deep. :slightly_smiling_face:

**Comment by ashishmagar600** [comment]
Second code snip worked. The first was throwing same error : 
`error: ‘mutex’ is not a member of ‘std’`
I checked it up and found this [link](https://stackoverflow.com/questions/14191566/c-mutex-in-namespace-std-does-not-name-a-type)

From answer I followed : 
`$ sudo update-alternatives --config x86_64-w64-mingw32-g++`
`<choose x86_64-w64-mingw32-g++-posix from the list>`

I again compiled and tested on windows, but having error : 
`libwinpthread.dll is missing` 
This was due to the exe again looking for libraries. So I statically linked and compiled again with ` -static ` 

The basic code I wrote, now is compiled on Linux (with mingw) and works on Windows.

Thanks for your help
